enhancements/scheduling/prevent-user-workloads-from-being-scheduled-on-control-plane-nodes.md

```diff
@@ -0,0 +1,737 @@
+    will show >1s latency and alert `WebhookAdmissionLatencyHigh` will fire.
+
+- disable the API extension (e.g. remove MutatingWebhookConfiguration `xyz`, remove APIService `foo`)
+
+  - What consequences does it have on the cluster health?
+
+    Examples:
+    - Garbage collection in kube-controller-manager will stop working.
+    - Quota will be wrongly computed.
+    - Disabling/removing the CRD is not possible without removing the CR instances. Customer will lose data.
+      Disabling the conversion webhook will break garbage collection.
+
+  - What consequences does it have on existing, running workloads?
+
+    Examples:
+    - New namespaces won't get the finalizer "xyz" and hence might leak resource X
+      when deleted.
+    - SDN pod-to-pod routing will stop updating, potentially breaking pod-to-pod
+      communication after some minutes.
+
+  - What consequences does it have for newly created workloads?
+
+    Examples:
+    - New pods in namespace with Istio support will not get sidecars injected, breaking
+      their networking.
+
+- Does functionality fail gracefully and will work resume when re-enabled without risking
+  consistency?
+
+  Examples:
+  - The mutating admission webhook "xyz" has FailPolicy=Ignore and hence
+    will not block the creation or updates on objects when it fails. When the
+    webhook comes back online, there is a controller reconciling all objects, applying
+    labels that were not applied during admission webhook downtime.
+  - Namespaces deletion will not delete all objects in etcd, leading to zombie
+    objects when another namespace with the same name is created. -->
+
+## Implementation History
+
+TODO
+
+<!-- Major milestones in the life cycle of a proposal should be tracked in `Implementation
+History`. -->
+
+## Alternatives
+
+### Have control-plane nodes tainted with NoExecute taint
+
+- Kubelet recognizes any NoExecute taint and rejects any pod that does not tolerate the taint. E.g. `node-role.kubernetes.io/master:NoExecute`.
+- The taint is a scheduler-free taint, i.e. no scheduler needs to understand taints and tolerations.
+- Any pod expected to be scheduled to a control plane node needs to tolerate the taint.
+- The admission plugin needs to reject any pod that is not allowed to tolerate the NoExecute taint.
+- When `masterSchedulable` is set to `true`, the admission plugin is disabled.
+- A scheduling group will no longer need a label selector. Instead, a list of taints will be required.
+- Each group will still need to keep a list of authorized service accouts and users
+- There's probably no guarantee for all control plane pods getting updated first (with the NoExecute toleration) before control plane nodes get tainted during an OCP upgrade. For that, all control plane pods need to tolerate the taint first in OCP X so nodes can be tainted in OCP X+1 without any unnecessary rejections/disruptions.
+
+### Validating Admission Policy
```

Comment:
Tired this as follows.

Firts tried creating a ValidatingAdmissionPolicy as described:

ValidatingAdmissionPolicy.yaml
```yaml
apiVersion: admissionregistration.k8s.io/v1beta1
kind: ValidatingAdmissionPolicy
metadata:
  name: "control-plane-scheduling-policy"
spec:
  failurePolicy: Fail
  matchConstraints:
    resourceRules:
    - apiGroups:   [""]
      apiVersions: ["v1"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["pods"]
  validations:
    - expression: "object.spec.tolerations.all(toleration, !(toleration.key == 'node-role.kubernetes.io/master' && toleration.effect == 'NoExecute'))"
```

And of course could not create it immediately as this is a beta feature and it is not enabled.

 ```
 error: resource mapping not found for name: "control-plane-scheduling-policy" namespace: "" from "ValidatingAdmissionPolicy.yaml": no matches for kind "ValidatingAdmi
ssionPolicy" in version "admissionregistration.k8s.io/v1beta1"                                                                                                        
ensure CRDs are installed first
 ```
 
 Checked  featuregates:
 
 ```
 oc edit featuregate cluster
 ```
 and changed it to enable TechPreview
 
 ```
 spec:
  featureSet: TechPreviewNoUpgrade
 ```
 
 Waited and `oc get featuregate cluster -o yaml`
 
 ```
 [...]
 status:
  featureGates:
  [...]
    enabled:
    [...]
      - name: ValidatingAdmissionPolicy
 ```
---

Then I am able to create and apply both the ValidatingAdmissionPolicy.yaml and ValidatingAdmissionPolicyBinding.yaml. (If you are still not able to created it, delete kube-apiserver pods so you force them to restart)

ValidatingAdmissionPolicy.yaml
```yaml
apiVersion: admissionregistration.k8s.io/v1beta1
kind: ValidatingAdmissionPolicy
metadata:
  name: "control-plane-scheduling-policy"
spec:
  failurePolicy: Fail
  matchConstraints:
    resourceRules:
    - apiGroups:   [""]
      apiVersions: ["v1"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["pods"]
  validations:
    -
```
ValidatingAdmissionPolicyBinding.yaml
```yaml
apiVersion: admissionregistration.k8s.io/v1beta1
kind: ValidatingAdmissionPolicyBinding
metadata:
  name: "control-plane-scheduling-policy-binding"
spec:
  policyName: "control-plane-scheduling-policy"
  validationActions: [Deny]
  matchResources:
    namespaceSelector:
      matchExpressions:
      - key: openshift.io/control-plane-namespace
        operator: DoesNotExist

```

---

Tainted a node directly:

```
kubectl taint nodes ip-10-0-79-111.ec2.internal node-role.kubernetes.io/master:NoExecute-
```

Tried scheduling pods directly in that node with nodeName set:

```
apiVersion: v1
kind: Pod
metadata:
  name: my-pod
  labels:
    app: my-app
spec:
  containers:
  - name: my-container
    image: nginx
    ports:
    - containerPort: 80
  nodeName: ip-10-0-79-111.ec2.internal
```

Pod does't even show in `k get pods`.

And kubelet logs on that nodes shows it is droped:

```
Mar 11 10:48:11.400517 ip-10-0-79-111 kubenswrapper[2129]: I0311 10:48:11.398322 2129 predicate.go:147] "Predicate failed on Pod" pod="default/my-pod" err="Predicate TaintToleration failed"
Mar 11 10:48:11.407547 ip-10-0-79-111 kubenswrapper[2129]: I0311 10:48:11.407525 2129 kubelet.go:2410] "SyncLoop UPDATE" source="api" pods=["default/my-pod"]
Mar 11 10:48:11.448263 ip-10-0-79-111 kubenswrapper[2129]: I0311 10:48:11.448238 2129 kubelet.go:2419] "SyncLoop DELETE" source="api" pods=["default/my-pod"]
Mar 11 10:48:11.451504 ip-10-0-79-111 kubenswrapper[2129]: I0311 10:48:11.451471 2129 kubelet.go:2413] "SyncLoop REMOVE" source="api" pods=["default/my-pod"]
Mar 11 10:50:50.070924 ip-10-0-79-111 kubenswrapper[2129]: I0311 10:50:50.070890 2129 kubelet.go:2403] "SyncLoop ADD" source="api" pods=["default/my-pod"]
Mar 11 10:50:50.071501 ip-10-0-79-111 kubenswrapper[2129]: I0311 10:50:50.070946 2129 topology_manager.go:215] "Topology Admit Handler" podUID="6c34e10a-08c8-4b8d-b59b-9d224ad7d112" podNamespace="default" podName="my-pod"
Mar 11 10:50:50.071501 ip-10-0-79-111 kubenswrapper[2129]: I0311 10:50:50.071118 2129 predicate.go:147] "Predicate failed on Pod" pod="default/my-pod" err="Predicate TaintToleration failed"
Mar 11 10:50:50.082297 ip-10-0-79-111 kubenswrapper[2129]: I0311 10:50:50.082261 2129 kubelet.go:2410] "SyncLoop UPDATE" source="api" pods=["default/my-pod"]
Mar 11 10:50:50.104118 ip-10-0-79-111 kubenswrapper[2129]: I0311 10:50:50.104073 2129 kubelet.go:2419] "SyncLoop DELETE" source="api" pods=["default/my-pod"]
Mar 11 10:50:51.208602 ip-10-0-79-111 kubenswrapper[2129]: I0311 10:50:51.208566 2129 kubelet.go:2419] "SyncLoop DELETE" source="api" pods=["default/my-pod"]
Mar 11 10:50:51.211972 ip-10-0-79-111 kubenswrapper[2129]: I0311 10:50:51.211948 2129 kubelet.go:2413] "SyncLoop REMOVE" source="api" pods=["default/my-pod"]
Mar 11 10:57:34.655752 ip-10-0-79-111 kubenswrapper[2129]: I0311 10:57:34.655667 2129 kubelet.go:2403] "SyncLoop ADD" source="api" pods=["default/my-pod"]
Mar 11 10:57:34.655752 ip-10-0-79-111 kubenswrapper[2129]: I0311 10:57:34.655714 2129 topology_manager.go:215] "Topology Admit Handler" podUID="890e0989-3179-42d1-a973-82e58242d5f7" podNamespace="default" podName="my-pod"
```

---

Tried setting nodeName and also tolerating the taint directly:

```
apiVersion: v1
kind: Pod
metadata:
  name: my-pod
  labels:
    app: my-app
spec:
  containers:
  - name: my-container
    image: nginx
    ports:
    - containerPort: 80
  nodeName: ip-10-0-79-111.ec2.internal
  tolerations:
  - key: "node-role.kubernetes.io/master"
    operator: "Exists"
    effect: "NoExecute"
```

CEL ValidatingAdmissionPolicy kicks in and does not let this happen, as the 'default' namespace is not allowed to receive pods with this toleration (with no correct labels).

```
The pods "my-pod" is invalid: : ValidatingAdmissionPolicy 'control-plane-scheduling-policy' with binding 'control-plane-scheduling-policy-binding' denied request: failed expression: object.spec.tolerations.all(toleration, !(toleration.key == 'node-role.kubernetes.io/master' && toleration.effect == 'NoExecute'))
```

---

After we label the default namespace with `kubectl label namespace default openshift.io/control-plane-namespace=true` we are finnaly able to apply the pod manifest that sets nodeName and has the tolaration for NoExecute taint